### ISSUE TITLE
Missed optimization opportunity: plain `for` loop slower than `foreach` or equivalent range pipeline

### ISSUE THREAD

**Issue by ZombineDev** [ldc-developers/ldc#2271]
#### LDC version:

```sh
$ ldc2 --version | head -n5
LDC - the LLVM D compiler (1.3.0):
  based on DMD v2.073.2 and LLVM 4.0.0
  built with LDC - the LLVM D compiler (1.3.0)
  Default target: x86_64-unknown-linux-gnu
  Host CPU: ivybridge
```

##### `sum0`:

```D
ulong sum0(ulong n)
{
    ulong res;

    for (ulong i = 1; i <= n; i++)
        res += i;

    return res;
}
```
```nasm
ulong example.sum0(ulong):
  xor eax, eax
  test rdi, rdi
  je .LBB3_3
  mov ecx, 1
.LBB3_2:     # This loop should be optimized away to a O(1) expression:
  add rax, rcx
  inc rcx
  cmp rcx, rdi
  jbe .LBB3_2
.LBB3_3:
  ret
```

##### `sum1`:

```D
ulong sum1(ulong n)
{
    ulong res;

    foreach (i; 1 .. n + 1)
        res += i;

    return res;
}
```

```nasm
# No loops - awesome!
ulong example.sum1(ulong):
  lea rax, [rdi + 1]
  cmp rax, 2
  jb .LBB4_1
  lea rax, [rdi - 1]
  lea rcx, [rdi - 2]
  mul rcx
  shld rdx, rax, 63
  lea rax, [rdx + 2*rdi - 1]
  ret
.LBB4_1:
  xor eax, eax
  ret
```

The above `sum1` generates very similar code as:

##### `sum2`:

```D
ulong sum2(ulong n)
{
    return (n * (n + 1)) / 2;
}
```

```nasm
ulong example.sum2(ulong):
  lea rax, [rdi + 1]
  imul rax, rdi
  shr rax
  ret
```

##### `sum1(uint)` and `sum2(uint)`:

The similarities between `sum1` and `sum2`, increase if we replace `ulong` with `uint`:

```nasm
uint example.sum1(uint):
  lea ecx, [rdi + 1]
  xor eax, eax
  cmp ecx, 2
  jb .LBB6_2
  lea eax, [rdi - 1]
  lea ecx, [rdi - 2]
  imul rcx, rax
  shr rcx
  lea eax, [rcx + 2*rdi - 1]
.LBB6_2:
  ret
```
```nasm
uint example.sum2(uint):
  lea eax, [rdi + 1]
  imul eax, edi
  shr eax
  ret
```

And finally, the star of the show:

##### `sum3`:

```D
ulong sum3(ulong n)
{
    import std.algorithm.iteration : sum;
    import std.range : iota;
    return iota(1, n + 1).sum;
}
```

```nasm
ulong example.sum3(ulong):
  lea rax, [rdi + 1]
  cmp rax, 2
  jae .LBB8_2
  xor eax, eax
  ret
.LBB8_2:
  lea rax, [rdi - 1]
  lea rcx, [rdi - 2]
  mul rcx
  shld rdx, rax, 63
  lea rax, [rdx + 2*rdi - 1]
  ret
```

Which (pleasantly surprising me) looks almost like `sum1(ulong)`.

For `100_000_000` iterations, the timings are:

| Function      | Time for `foreach (_; 0 .. iterations) global += func(input)` |
|---------------|------------------------------------------------------------|
| `sum0(ulong)` | 44 secs, 670 ms, 33 μs, and 2 hnsecs                       |
| `sum1(ulong)` | 180 ms, 139 μs, and 9 hnsecs                               |
| `sum2(ulong)` | 154 ms and 595 μs                                          |
| `sum3(ulong)` | 231 ms, 670 μs, and 3 hnsecs                               |

So, the obvious question is: can all of the above functions be optimized like `sum2`?

Full source code: https://godbolt.org/g/NgnzP3

**Comment by kinke** [comment]
Changing sum0 to
```D
    ...
    ++n;
    for (ulong i = 1; i < n; i++)
    ...
```
makes it produce the folded version. I think we already have at least one similar issue by @JohanEngelen. May even have been LLVM specific.

**Comment by kinke** [comment]
That makes the codegen for `sum0` and `sum1` identical; `sum3` diverges only in that it performs a jump if the range isn't empty, while sum0 and sum1 only jump if it *is* empty.

**Comment by JohanEngelen** [comment]
@kinke is right. Because `n+1` rolls over for `n == ulong.max`, `sum0()` and `sum1()` are not identical semantically, so they cannot result in the same code.
`sum0()` is not optimized further probably because it has to result in an infinite loop for `n == ulong.max`.
@ZombineDev Can you redo the analysis, taking rollover and `n == ulong.max` into account?

@kinke The similar issue you mean is perhaps the `+2` loops? https://github.com/ldc-developers/ldc/issues/2154

**Comment by kinke** [comment]
Yep, that's exactly the one I meant.

**Comment by ZombineDev** [comment]
Thanks for the info guys. Indeed one needs to be extra careful with integer overflow, infinite loops and undefined behavior in general.

Looking at #2154 it looks like it may be a problem with llvm. For example, this should should terminate, but the loop is still there:
```D
long sum0(long n)
{
    long res;

    for (long i = 1; i > 0 && i <= n; i++)
        res += i;

    return res;
}
```

Meanwhile, I noticed that clang 4.0 manages to optimize both `for (int64_t i = 1; i <= n; i++)` and `for (int64_t i = 1; i < n + 1; i++)` to the same O(1) assembly. Interestingly, for `for (int64_t i = 1; i > 0 && i <= n; i++)` it does something different, though still O(1) - https://godbolt.org/g/rw11dj

(Other C++ compilers that I tested (CL, ICC, GCC) miss this optimization and do some sort of loop unrolling and in the case of ICC heavy auto-vectorization.)

Could there be some extra optimization that clang applies before llvm, taking advantage of undefined behavior?

Is there a way to tell the LLVM IR, that e.g. `n` will always be e.g. `< long.max / 2`, so that it could do extra optimizations?

**Comment by kinke** [comment]
> Meanwhile, I noticed that clang 4.0 manages to optimize both for (int64_t i = 1; i <= n; i++)

Try using `uint64_t`, it behaves identically to LDC with D `ulong` then.

**Comment by JohanEngelen** [comment]
@ZombineDev Important: signed overflow is UB in C and C++ .

**Comment by ZombineDev** [comment]
Right, forgot that C++ doesn't mandate two's complement signed integer representation.
What about this, though: `for (long i = 1; i > 0 && i <= n; i++)` with LDC? Is there some UB I am missing?

**Comment by kinke** [comment]
> The optimization is still there with clang for the unsigned variant: for (uint64_t i = 1; i != 0 && i <= n; i++)

Only if `i != 0` is included in the condition; if I add that for LDC as well, it's also folded. ;)

**Comment by JohanEngelen** [comment]
>  Is there some UB I am missing?

Perhaps because things are _not_ UB in D with `long`, correct results must be calculated when n is large and then the constant folded optimization is not valid (`res` will be overflowing earlier than `i`, so it's a little hard to reason about this). You could try and run the code with n = large and check whether the output is the same or not for the ` i != 0` and the `i > 0` versions. 
Interesting stuff for a blog post btw ;-)
